src/ispc/kernels/lanczos3.ispc

```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
+        else
+            clean_and_write_4_channels(color4, scratch_write_address, scratch_space);
     }
-    col /= weight;
-    return col * 255;
-}
 
-export void resample(uniform uint32 width, uniform uint32 height, uniform uint32 stride, uniform uint8 num_channels, uniform uint32 target_width, uniform uint32 target_height, uniform const uint8 src_data[], uniform uint8 out_data[]) {
-    uniform Image src = {src_data, {width, height}};
-    uniform float<2> target_size = {(float)target_width, (float)target_height};
-    uniform float<2> inv_target_size = {1.0 / target_width, 1.0 / target_height};
+    // Accumulate the scratch space data along the height
+    // Downsamples the src_height X target_width image to target_height * target_width
+    foreach_tiled(y =  0 ... target_height, x = 0 ... target_width) {
+
+        uint32 src_height_start = vertical_weight_collection->starts[y];
+        uint32 num_vertical_weights = vertical_weight_collection->weight_counts[y];
+        float* vertical_weights = vertical_weight_collection->values[y];
 
-    foreach_tiled (y = 0 ... target_height, x = 0 ... target_width) {
-        float yf = y / target_size.y;
-        float xf = x / target_size.x;
+        float<3> color3 = {0.0f, 0.0f, 0.0f};
+        float<4> color4 = {0.0f, 0.0f, 0.0f, 0.0f};
+        for (uint32 i = 0; i < num_vertical_weights; i++) {
+            float weight = vertical_weights[i];
+            uint32 scratch_y = src_height_start + i;
+            uint64 scratch_read_address = (scratch_y * target_width + x) * num_channels;
 
-        float<2> uv = { xf, yf };
+            uniform uint8<3>* varying scratch_pixel_ptr = (uniform uint8<3>* varying)(scratch_space + scratch_read_address);
+            uint8<3> scratch_color = *scratch_pixel_ptr;
+
+            if (num_channels == 3)
+                color3 += sample_3_channels(scratch_read_address, scratch_space) * weight;
+            else
+                color4 += sample_4_channels(scratch_read_address, scratch_space) * weight;
+        }
 
-        uint8<4> s = resample_internal(src, uv, target_size, inv_target_size, num_channels);
+        uint64 out_write_address = (y * target_height + x) * num_channels;
 
-        for (uniform int i = 0; i < num_channels; i++)
-            out_data[(x + y * target_width) * num_channels + i] = s[i];
+        if (num_channels == 3)
+            clean_and_write_3_channels(color3, out_write_address, out_data);
+        else
+            clean_and_write_4_channels(color4, out_write_address, out_data);
     }
 }
+
+/// scratch_space must be at least src_height * target_width pixels big
+export void resample_with_cache_3(uniform uint32 src_width, uniform uint32 src_height, uniform uint32 target_width, uniform uint32 target_height,
+    uniform const Cache * uniform cache, uniform uint8 scratch_space[], uniform const uint8 src_data[], uniform uint8 out_data[]) {
+    uniform const uint32 num_channels = 3;
```

Comment:
Remind me: did it make a difference to pass this as arugment from Rust? (I hope it does, the compiler should be able to deadstrip the if/else for 3/4 channels after const-folding)